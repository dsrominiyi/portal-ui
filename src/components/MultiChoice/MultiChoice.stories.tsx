import React from 'react';
import { storiesOf } from '@storybook/react';

import MultiChoice from '.';

const stories = storiesOf('MultiChoice', module);

export const render = (
  options: string[],
  selectedOption?: string,
  multiSelect?: boolean,
  selectedOptions?: string[],
  disabledOptions?: string[]
) => (props?: any) => (
  <MultiChoice
    options={options}
    onClick={(props && props.onClick) || (() => null)}
    selectedOption={selectedOption}
    multiSelect={multiSelect}
    selectedOptions={selectedOptions}
    disabledOptions={disabledOptions}
  />
);

export const options = ['Car Park', 'Garage', 'Road', 'Off-road'];

export const storyRenders = {
  notSelected: render(options),
  selected: render(options, options[1]),
  multiSelect: render(options, undefined, true, [options[0], options[2]]),
  disabledOptions: render(options, undefined, true, [], [options[0], options[2]])
};

stories.add('Not selected', storyRenders.notSelected);
stories.add('Selected', storyRenders.selected);
stories.add('Multi-select', storyRenders.multiSelect);
stories.add('Disabled options', storyRenders.disabledOptions);
