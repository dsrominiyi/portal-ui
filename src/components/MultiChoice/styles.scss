.options {
  > * {
    display: inline-flex;
    margin: 0.5rem;
  }
}
