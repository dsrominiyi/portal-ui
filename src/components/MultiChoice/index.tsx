import React, { FunctionComponent } from 'react';
import QuestionButton from '../Button/QuestionButton';
import styles from './styles.scss';

export interface Props {
  options: string[];
  selectedOption?: string;
  multiSelect?: boolean;
  selectedOptions?: string[];
  disabledOptions?: string[];
  attributes?: { [attr: string]: string };
  onClick(option: string): void;
}

export const updateSelectedOptions = (selectedOptions: string[], option: string) => {
  selectedOptions.includes(option)
    ? selectedOptions.splice(selectedOptions.indexOf(option), 1)
    : selectedOptions.push(option);

  return selectedOptions;
};

export const MultiChoice: FunctionComponent<Props> = ({
  options,
  selectedOption,
  multiSelect,
  selectedOptions = [],
  disabledOptions = [],
  attributes = {},
  onClick
}: Props) => {
  const testId = attributes['test-id'];
  const renderedOptions = options.map((option, index) => (
    <QuestionButton
      key={index}
      label={option}
      isActive={multiSelect ? selectedOptions.includes(option) : option === selectedOption}
      isDisabled={disabledOptions.includes(option)}
      onClick={() => onClick(option)}
      attributes={{ 'test-id': testId ? `${testId}_option-${index + 1}` : '' }}
    />
  ));
  return (
    <div className={styles.options} {...attributes}>
      {renderedOptions}
    </div>
  );
};

MultiChoice.displayName = 'MultiChoice';
export default MultiChoice;
