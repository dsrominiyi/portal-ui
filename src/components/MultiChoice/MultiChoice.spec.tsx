import { shallow } from 'enzyme';

import { storyRenders, options } from './MultiChoice.stories';
import testStorySnapshots from '../../helpers/test/testStorySnapshots';

describe('MultiChoice', () => {
  test('component story renders match snapshots', () => {
    testStorySnapshots(storyRenders);
  });

  test('onClick is called with the selected option', () => {
    const onClick = jest.fn();
    const MultiChoice = shallow(storyRenders.notSelected({ onClick }));
    const buttons = MultiChoice.find('QuestionButton');

    buttons.forEach((button, index) => {
      button.simulate('click');

      expect(onClick).toHaveBeenCalledWith(options[index]);
    });
  });
});
