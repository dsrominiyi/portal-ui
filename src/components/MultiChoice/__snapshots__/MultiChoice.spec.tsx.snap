// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiChoice component story renders match snapshots: disabledOptions 1`] = `
<div
  className="options"
>
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={true}
    key="0"
    label="Car Park"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={false}
    key="1"
    label="Garage"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={true}
    key="2"
    label="Road"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={false}
    key="3"
    label="Off-road"
    onClick={[Function]}
  />
</div>
`;

exports[`MultiChoice component story renders match snapshots: multiSelect 1`] = `
<div
  className="options"
>
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={true}
    isDisabled={false}
    key="0"
    label="Car Park"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={false}
    key="1"
    label="Garage"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={true}
    isDisabled={false}
    key="2"
    label="Road"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={false}
    key="3"
    label="Off-road"
    onClick={[Function]}
  />
</div>
`;

exports[`MultiChoice component story renders match snapshots: notSelected 1`] = `
<div
  className="options"
>
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={false}
    key="0"
    label="Car Park"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={false}
    key="1"
    label="Garage"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={false}
    key="2"
    label="Road"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={false}
    key="3"
    label="Off-road"
    onClick={[Function]}
  />
</div>
`;

exports[`MultiChoice component story renders match snapshots: selected 1`] = `
<div
  className="options"
>
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={false}
    key="0"
    label="Car Park"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={true}
    isDisabled={false}
    key="1"
    label="Garage"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={false}
    key="2"
    label="Road"
    onClick={[Function]}
  />
  <QuestionButton
    attributes={
      Object {
        "test-id": "",
      }
    }
    isActive={false}
    isDisabled={false}
    key="3"
    label="Off-road"
    onClick={[Function]}
  />
</div>
`;
